### PR TITLE
Carousel: Add option to vary display interval on a per-slide basis

### PR DESCRIPTION
**Pull request by bassjobsen**
Add the option to set the interval per slide leveraging the `data-interval` attribute, see also http://stackoverflow.com/questions/27047620/custom-carousel-intervals/27049608

### CODE REVIEW

**Comment by cvrebert** [comment]
@fat Ping.

File: js/carousel.js [hunk 1]
```diff
@@ -59,7 +60,7 @@
 
     this.options.interval
       && !this.paused
-      && (this.interval = setInterval($.proxy(this.next, this), this.options.interval))
+      && (this.interval = setInterval($.proxy(this.next, this), this.duration = this.$element.find('.item.active').data('duration') || this.options.interval))
```
Comment (hnrch02) [inline js/carousel.js]: Any particular reason for using `this.$element.find('.item.active')` instead of `this.$active` here? And also, why save the value. I don't see it being used anywhere else.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 4db2c120dd3cd76a39f8eb565543250ffb4ad9a3
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52811662

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by bassjobsen** [comment]
@hnrch02 thanks. You are right there is no reason to save `this.duration`. Although i used this value in the unit test before. In fact the unit should test the value set for `setInterval` (replace with `setTimeout` in my last update). I don't know how to test that value? Now the test seems to test more reading of data attributes instead of durations set.

You also ask if i could use `this.$active`, i don't think so (or at least not in the current code). The `cycle` function will be the first called when initialising the plugin without any parameters. In this situation `this.$active` has not been set yet.

You can also initialise the plugin with `.carousel(number)` or `.carousel('next')`. `.carousel(number)` set  `this.$active` indeed, but  `.carousel(number)` set a [local value for $active](https://github.com/twbs/bootstrap/blob/master/js/carousel.js#L118). I also wondered if both calls make sense when the carousal has not been initialized before? But when not setting `$active` in the `slide` function let directly fail the first test "should not fire slid when slide is prevented'".

I think `this.$active` should set on initialisation always (and not per function). I will try to fix this in my next update.

I also found the [usage of setInterval](https://github.com/twbs/bootstrap/blob/master/js/carousel.js#L62) as far as in understand, `setInterval` fires immediately and then after the set interval. The first immediately trigger seems to be blocked by the running transition (i don't know what happens when transitions are not supported, but probably not the right thing), `setTimeout` seems to better fit?

And then it seems [cycle](https://github.com/twbs/bootstrap/blob/master/js/carousel.js#L167) is already called before the transition has ended. The active class move in the transition, so cycle runs on the previous active item instead of the current. I realized now that in my current code the interval become (unexpected) set interval + transition duration.

File: less/component-animations.less [hunk 2]
```diff
@@ -31,4 +31,9 @@
   .transition-property(~"height, visibility");
   .transition-duration(.35s);
   .transition-timing-function(ease);
+  &.width {
```
Comment (cvrebert) [inline less/component-animations.less]: This change seems unrelated. Can you revert it?

File: js/collapse.js [hunk 3]
```diff
@@ -102,7 +102,7 @@
 
     var dimension = this.dimension()
 
-    this.$element[dimension](this.$element[dimension]())[0].offsetHeight
+    this.$element[dimension](this.$element[dimension]())[0][$.camelCase(['inner', dimension].join(''))]
```
Comment (cvrebert) [inline js/collapse.js]: Please revert this unrelated change.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 0664f602b5232e67b90fca7a97e336387b7d362b
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52824540

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
I don't understand why you also have `duration` stuff in here. The transition duration also involves the CSS and thus can't be easily customized, unlike the interval, which is what this PR is supposed to be about.

**Comment by bassjobsen** [comment]
@cvrebert do you refer to the `.emulateTransitionEnd(Math.min(Carousel.TRANSITION_DURATION,($active.data('duration') || this.interval)))`. I think you are right about that. But i should possible check the situation that duration < Carousel.TRANSITION_DURATION.

Now (after the PR) it runs: duration first slide + Carousel.TRANSITION_DURATION, and so on.

**Comment by cvrebert** [comment]
Yes, that's what I'm referring to, in part.
Also, `this.$element.find('.item.active').data('duration')` ought to be `this.$element.find('.item.active').data('interval')` to avoid confusion.

**Comment by mdo** [comment]
I'd like to punt on this for v3. We're not adding any other features and this falls under that banner. Plus, I don't see a super strong benefit to varying the duration right now. Appreciate the effort though!
